### PR TITLE
[IMP] formulas: add CONVERT formula

### PR DESCRIPTION
**Pull request by anhe-odoo**
## Description:

TODO

Task: : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Comment by rrahir** [comment]
"... but he never did add the tests"
Jokes aside I like the idea of that formula - it might be very useful in the global roadmap of the company ^^

File: tests/functions/module_parser.test.ts [hunk 1]
```diff
@@ -0,0 +1,262 @@
+import { evaluateCell } from "../test_helpers/helpers";
+
+describe("CONVERT formula", () => {
+  test("Arguments validation", () => {
+    expect(evaluateCell("A1", { A1: "=CONVERT()" })).toBe("#BAD_EXPR");
+    expect(evaluateCell("A1", { A1: "=CONVERT(1)" })).toBe("#BAD_EXPR");
+    expect(evaluateCell("A1", { A1: '=CONVERT(1,"m")' })).toBe("#BAD_EXPR");
+    expect(evaluateCell("A1", { A1: '=CONVERT(1,,"m")' })).toBe("#ERROR"); // invalid unit
+    //expect(evaluateCell("A1", { A1: '=CONVERT(1,"unknown","m")' })).toBe("#ERROR"); // invalid unit
```
Comment (hokolomopo) [inline tests/functions/module_parser.test.ts:9]: comment

File: tests/functions/module_parser.test.ts [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    ["psi", "Pa", "1", 6894.76],
+    ["atm", "Pa", "1", 101325],
+  ])("correctly convert pressure unit (%s to %s)", (from, to, value, result) => {
+    expect(evaluateCell("A1", { A1: `=CONVERT(A2, "${from}", "${to}")`, A2: value })).toBeCloseTo(
+      result
+    );
+    expect(evaluateCell("A1", { A1: `=CONVERT(1, "${from}", "${from}")` })).toBeCloseTo(1);
+    expect(
+      evaluateCell("A1", { A1: `=CONVERT(CONVERT(1, "${from}", "${to}"), "${to}", "${from}")` })
+    ).toBeCloseTo(1);
```
Comment (hokolomopo) [inline tests/functions/module_parser.test.ts:135]:
```suggestion
    expect(evaluateCell("A1", { A1: `=CONVERT(1, "${from}", "${from}")` })).toBeCloseTo(1);
    expect(
      evaluateCell("A2", { A2: `=CONVERT(A1, "${to}", "${from}")` })
    ).toBeCloseTo(1);
```
nitpick, a bit shorter

**Review by hokolomopo — COMMENTED** [review]
Damn you really implemented all the units, that's impressive :exploding_head: 

The commit message/PR message seems a bit messed up tho :)

File: tests/functions/module_parser.test.ts (hunk 2 above)
Comment (anhe-odoo) [reply tests/functions/module_parser.test.ts:135]: Nope

File: src/functions/module_parser.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          end_conversion.category
+        ),
+      };
+    }
+    return {
+      value:
+        end_conversion.inverseTransform(
+          start_conversion.factor * start_conversion.transform(_value)
+        ) / end_conversion.factor,
+      format: value?.format || end_unit?.format || start_unit?.format,
```
Comment (LucasLefevre) [inline src/functions/module_parser.ts:55]: does it really makes sense to get the format from the units ?
```suggestion
      format: value?.format,
```

File: src/functions/module_parser.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        message: _t("Invalid units of measure ('%s')", _end_unit),
+      };
+    }
+    if (start_conversion.category !== end_conversion.category) {
+      return {
+        value: CellErrorType.GenericError,
+        message: _t(
+          "Incompatible units of measure ('%s' vs '%s')",
+          start_conversion.category,
+          end_conversion.category
```
Comment (LucasLefevre) [inline src/functions/module_parser.ts:46]: categories are always displayed in English here. They should be translated right ?

**Review by LucasLefevre — COMMENTED** [review]
nitpicking mainly. 
I did not check if conversions are correct, I trust you :wink:

File: src/functions/module_parser.ts (hunk 3 above)
Comment (anhe-odoo) [reply src/functions/module_parser.ts:55]: No, you're right :)

File: src/functions/module_parser.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        message: _t("Invalid units of measure ('%s')", _endUnit),
+      };
+    }
+    if (startConversion.category !== endConversion.category) {
+      return {
+        value: CellErrorType.GenericError,
+        message: _t(
+          "Incompatible units of measure ('%s' vs '%s')",
+          _t(startConversion.category),
+          _t(endConversion.category)
```
Comment (LucasLefevre) [inline src/functions/module_parser.ts:46]: This doesn't work. It only works on static strings :wink: 

We could add a mapping.

```js
const blabla = {
    volume: _t("Volume"),
    ...
}
```

File: src/functions/helper_parser.ts [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  Pi: Math.pow(2, 50),
+  Ti: Math.pow(2, 40),
+  Gi: Math.pow(2, 30),
+  Mi: Math.pow(2, 20),
+  ki: Math.pow(2, 10),
+};
+
+const TRANSLATED_CATEGORIES = Object.assign(
+  {},
+  ...Array.from(new Set(Object.values(UNITS).map((u) => u.category)), (u) => ({ [u]: _t(u) }))
```
Comment (LucasLefevre) [inline src/functions/helper_parser.ts:259]: still doesn't work. There can never be variables inside `_t(...)`, only hard-coded static strings.
We need an explicit and hard-coded mapping

```js
const TRANSLATED_CATEGORIES = {
    volume: _t("Volume"),
    ...
}
```

**Review by LucasLefevre — APPROVED** [review]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo r+

**Pull request closed without merging**